plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergBucketFunction.java

```diff
@@ -0,0 +1,140 @@
+import static io.trino.spi.function.InvocationConvention.InvocationReturnConvention.FAIL_ON_NULL;
+import static io.trino.spi.function.InvocationConvention.simpleConvention;
+import static io.trino.spi.type.TypeUtils.NULL_HASH_CODE;
+import static java.util.Objects.requireNonNull;
+
+public class IcebergBucketFunction
+        implements BucketFunction
+{
+    private final int bucketCount;
+
+    private final List<PartitionColumn> partitionColumns;
+    private final List<MethodHandle> hashCodeInvokers;
+
+    public IcebergBucketFunction(
+            TypeOperators typeOperators,
+            PartitionSpec partitionSpec,
+            List<IcebergColumnHandle> partitioningColumns,
+            int bucketCount)
+    {
+        requireNonNull(partitionSpec, "partitionSpec is null");
+        checkArgument(!partitionSpec.isUnpartitioned(), "empty partitionSpec");
+        checkArgument(bucketCount > 0, "Invalid bucketCount: %s", bucketCount);
+        this.bucketCount = bucketCount;
+
+        Map<Integer, Integer> fieldIdToInputChannel = new HashMap<>();
+        for (int i = 0; i < partitioningColumns.size(); i++) {
+            fieldIdToInputChannel.put(partitioningColumns.get(i).getId(), i);
+        }
+        partitionColumns = partitionSpec.fields().stream()
+                .map(field -> {
+                    Integer channel = fieldIdToInputChannel.get(field.sourceId());
+                    checkArgument(channel != null, "partition field not found: %s", field);
+                    Type inputType = partitioningColumns.get(channel).getType();
+                    ColumnTransform transform = getColumnTransform(field, inputType);
+                    return new PartitionColumn(channel, transform.getValueTransform(), transform.getType());
+                })
+                .collect(toImmutableList());
+        hashCodeInvokers = partitionColumns.stream()
+                .map(PartitionColumn::getResultType)
+                .map(type -> typeOperators.getHashCodeOperator(type, simpleConvention(FAIL_ON_NULL, NEVER_NULL)))
+                .collect(toImmutableList());
+    }
+
+    @Override
+    public int getBucket(Page page, int position)
+    {
+        long hash = 0;
+
+        for (int i = 0; i < partitionColumns.size(); i++) {
+            PartitionColumn partitionColumn = partitionColumns.get(i);
+            Block block = page.getBlock(partitionColumn.getSourceChannel());
+            Object value = partitionColumn.getValueTransform().apply(block, position);
+            long valueHash = hashValue(hashCodeInvokers.get(i), value);
+            hash = (31 * hash) + valueHash;
+        }
+
+        return (int) ((hash & Long.MAX_VALUE) % bucketCount);
+    }
+
+    private static long hashValue(MethodHandle method, Object value)
+    {
+        if (value == null) {
+            return NULL_HASH_CODE;
+        }
+        try {
+            return (long) method.invoke(value);
```

Comment:
I followed what @raunaqmorarka did in `HivePartitionedBucketFunction` (https://github.com/trinodb/trino/pull/6866).
Looks like supporting this would be not a trivial change (`MethodGenerator` might want to become public, this may require making the API a bit more fool-proof, per it's assumptions about `type x interface` implicitly defining the operation). Also, a different strategy could be compiling the whole IcebergBucketFunction class.

all in all, it's hard to embark on such endeavours without having iceberg benchmarks (which we don't have yet)

@raunaqmorarka do you think this is something that could be addressed in a follow up?

Reply:
For this kind of change some targeted JMH benchmarking is probably enough. But I agree we can keep this for a follow-up, not necessary to address here.